### PR TITLE
Ml cleaning

### PR DESCRIPTION
**Pull request by XingdaCen**
New features:

1.  Avoid adding duplicate models if two same model flags are given.
2. Allow to change the resolution of the object detection networks (caffe-based and yolo).
3. Allow to change the confidence threshold of the object detection networks (caffe-based and yolo).
4. Allow to pass two model theme flags (`-outdoors and -indoors`) to set up object detection networks, its resolution and prioritiser. Currently the `-outdoors` flag uses `MobleNetSSD_deploy` model with resolution 300x300 while `-indoors` flag uses `no_bn` model with resolution 640x480. The two flags use the same prioritiser module `name_prio_smart`.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -381,3 +381,4 @@ FodyWeavers.xsd
 .DS_Store
 Intel_SighPP_CMake/.DS_Store
 Intel_SighPP_CMake/Intel_SighPP_CMake/.DS_Store
+Intel_SighPP_CMake/Intel_SighPP_CMake/.DS_Store
```
Comment (GisliGeorgs) [inline .gitignore]: the line above with .DS_Store shoud be enough, the files that are already tracked need to be removed from tracking by git to be removed 😄

File: Intel_SighPP_CMake/CMakeLists.txt [hunk 2]
```diff
@@ -3,18 +3,8 @@
 #
 cmake_minimum_required (VERSION 3.8)
 
-set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
-set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/CMakeLists.txt]: Idk much about CMake but should this be removed?

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp [hunk 3]
```diff
@@ -246,17 +279,16 @@ int main(int argc, char** argv)
 	smart_priority prio_smart(name_prio_smart);
 	SPDLOG_INFO("Using prioritiser module: {}", prio_smart.get_name());
 
-	//std::string name_prio_size = "size";
-	//priority_module* prio_depth = new size_priority(&name_prio_size);
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp]: cant this and the empty lines be removed?

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp [hunk 4]
```diff
@@ -192,7 +200,33 @@ int main(int argc, char** argv)
...
+				}
 			}
+			else if(next_arg.compare("-indoors") == 0)
+			{
+				theme = 2;
+				for(auto i = indoors_model.begin(); i != indoors_model.end(); ++i){
+					std::pair<std::set<string>::iterator,bool> ret = caffe_models_names.insert(*i);
+					if(ret.second){
+						CaffeModelImpl caffe_model = create_caffe_network(*i);
+						caffe_model.set_resolution(640,480);
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp:213]: Can't this be set in the constructor?

Comment (GisliGeorgs) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp:213]: Also remember that the realsense camera default setup is 640x480 if that matters for this purpose.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/ml_interface.hpp [hunk 5]
```diff
@@ -12,5 +12,11 @@ struct ModelInterface {
  	/// <param name="color_matrix">color matrix obtained from the camera</param>
  	/// <param name="depth_matrix">depth matrix obtained from the camera</param>
 	virtual ClassificationResult do_work(cv::Mat color_matrix, cv::Mat depth_matrix) = 0;
+	/// <summary>
+ 	/// Function to set frame's resolution in object detection networks
+ 	/// </summary>
+ 	/// <param name="width">width of the frame resolution</param>
+ 	/// <param name="height">height of the frame resolution</param>
+	virtual void set_resolution(size_t width, size_t height) = 0;
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/ml_interface.hpp]: I thinks rather belongs in the constructor than in a separate function

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/yolo_impl.cpp [hunk 6]
```diff
@@ -144,4 +144,8 @@ struct YoloModelImpl : public ModelInterface {
 		return classification_result;
 	}
 
+	void set_resolution(size_t width, size_t height) override {
```
Comment (GisliGeorgs) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/yolo_impl.cpp]: Same as with caffe, i feel this belongs in the constructor

**Review by GisliGeorgs — COMMENTED** [review]
Looks alright, look at comments, nothing major.

File: Intel_SighPP_CMake/CMakeLists.txt (hunk 2 above)
Comment (XingdaCen) [reply Intel_SighPP_CMake/CMakeLists.txt]: I don't know either. I think I'll just keep these two lines and make is the same as the one in the Development branch.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp (hunk 4 above)
Comment (XingdaCen) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp:213]: Yeah we can set it in the constructor but I think it is also acceptable to have a separate function. So in some cases you may need to use the same network but with different resolution so by using a separate function you don't have to initiate another one.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp [hunk 7]
```diff
@@ -246,17 +274,16 @@ int main(int argc, char** argv)
...
-	//add modules
 	prioritiser->add_module(prio_smart);
 	// Todo: load prio model from flag
-	prioritiser->set_module(name_prio_smart);
+	switch (theme)
+	{
+		case 1:
+		case 2:
+		default: prioritiser->set_module(name_prio_smart);
+	}
```
Comment (SvenStruanFinlay) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp:272]: This is good. I just realised i think we need to split this file into cpp hpp

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/yolo_impl.cpp [hunk 8]
```diff
@@ -17,9 +17,9 @@ struct YoloModelImpl : public ModelInterface {
...
 	cv::dnn::Net net;
 	vector<string> class_names;
+	size_t inWidth;
+	size_t inHeight;	
 	
-	const size_t inWidth = 640;
-	const size_t inHeight = 480;
 	const float WHRatio = inWidth / (float)inHeight;
 	const float inScaleFactor = 0.007843f;
 	const float meanVal = 127.5;
```
Comment (SvenStruanFinlay) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/yolo_impl.cpp]: I think these variables could be put into the interface? They are the same in both caffe and yolo?

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/yolo_impl.cpp [hunk 9]
```diff
@@ -28,21 +28,21 @@ struct YoloModelImpl : public ModelInterface {
...
-		
 		net = cv::dnn::readNetFromDarknet(modelConfiguration,modelWeights);
-		
 		net.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
 		// Use configuration to speed up the net.forward
 		net.setPreferableTarget(cv::dnn::DNN_TARGET_OPENCL);		
-		
 		class_names = read_class_name_file(class_names_path);
+		inWidth = 640;
+		inHeight = 480;
```
Comment (SvenStruanFinlay) [inline Intel_SighPP_CMake/Intel_SighPP_CMake/src/ml_lib/yolo_impl.cpp]: Again perhaps these variables could be in the super constructor. Not 100% needed though.

**Review by SvenStruanFinlay — COMMENTED** [review]
Looks good. Comments are good. Could be some reformating but not urgent.

File: Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp (hunk 7 above)
Comment (XingdaCen) [reply Intel_SighPP_CMake/Intel_SighPP_CMake/src/Intel_Sight_++.cpp:272]: Yeah we can do that if we have more time but i think we can just leave it for now.

**Review by GisliGeorgs — APPROVED** [review]
Looks alright.

**Review by lspcool — APPROVED** [review]

**Review by SvenStruanFinlay — APPROVED** [review]

**Pull request merged**